### PR TITLE
Add geographical information to the plant static attributes data frame

### PR DESCRIPTION
**Pull request by rouille**
### Purpose
Add the following fields to the plant static attributes: "latitude", "longitude", "county", "city", "zip_code", "street_address" and "capacity_mw". Closes CAR-4208.

Add missing typehints and add/format docstrings in modules that were edited.

### What the code is doing
* Create a function to load the geographical information of the plants from the raw EIA-860 data
* Load the `plants_entity_eia` PUDL table to retrieve the geographical information of the plants
* Supplement the PUDL table with data from the raw EIA-860 when entries are missing
* Output the plant static attributes file to disk for years < 2019

### Testing
Successfully ran the 2010 pipeline.

### Where to look
The `oge.load_data` module where the new function load the geographical information from the raw EIA-860 and the `oge.helpers` module where fields to the plant static attributes data frame are added.

### Usage Example/Visuals
Screen shot of the exported 2010 plant static attributes table
Screenshot 2024-05-21 at 1 25 23 AM

### Review estimate
15min

### Future work
CAR-4205

### Checklist
- [x] Update the documentation to reflect changes made in this PR
- [x] Format all updated python files using `black`
- [x] Clear outputs from all notebooks modified
- [x] Add docstrings and type hints to any new functions created

### CODE REVIEW

File: src/oge/data_pipeline.py
Comment (grgmiller) [inline src/oge/data_pipeline.py:697]: This should be `pd.NA` since shaped_plant_id should have an Int dtype

File: src/oge/helpers.py
Comment (grgmiller) [inline src/oge/helpers.py:184]: I'm assuming you removed this to have more consistent dtypes, but did you confirm that the original behavior is as expected? ie that missing BAs are getting filled properly?

File: src/oge/helpers.py
Comment (grgmiller) [inline src/oge/helpers.py:208]: should we just use `column_checks.get_dtypes()` here?

File: src/oge/helpers.py
Comment (grgmiller) [inline src/oge/helpers.py:220]: should we just use column_checks.get_dtypes() here?

File: src/oge/helpers.py
Comment (grgmiller) [inline src/oge/helpers.py:258]: To be consistent with the deleted code above, we should use `pd.NA` here

File: src/oge/helpers.py
Comment (grgmiller) [inline src/oge/helpers.py:409]: I would propose removing street address and zip code, but adding the `plant_name_eia` column.
I don't think we need street address since we have lat/long info

File: src/oge/helpers.py
Comment (grgmiller) [inline src/oge/helpers.py:415]: I get why you use latest_validated_year here, but do we download this file for the latest validated year no matter what year we're running? Just want to make sure we don't run into a filenotfounderror.

File: src/oge/helpers.py
Comment (grgmiller) [inline src/oge/helpers.py:309]: Why is rounding necessary? Do we ever have data with more than 2 decimals?
We may want to keep the original level of detail

File: src/oge/helpers.py
Comment (grgmiller) [inline src/oge/helpers.py:305]: Why do we use this query format here rather than just: `generators_capacity[generators_capacity["report_date"] == generators_capacity["report_date"].max()]]`

**Review by grgmiller — CHANGES_REQUESTED** [review]
See comments and requested changes. 

I'm proposing that we add the plant name to this file, and drop the street address/zip code unless you think we will use those.

File: src/oge/helpers.py
Comment (rouille) [reply src/oge/helpers.py:208]: `ba_name` is not in `get_types()`. Should I add it?

File: src/oge/helpers.py
Comment (rouille) [reply src/oge/helpers.py:220]: Same issue, `name` is not in `get_dtypes()`. This is more tricky since a key named name is ambiguous. I would change the name of the header of the first column in the **utility_name_ba_code_map.csv** file to `utility_name`.

File: src/oge/helpers.py
Comment (rouille) [reply src/oge/helpers.py:409]: Sounds good

File: src/oge/helpers.py
Comment (rouille) [reply src/oge/helpers.py:415]: This is not an issue when running the pipeline since the EIA-860 files are downloaded for the latest validated year in Step 1.

Now if you believe we could use this function outside of the pipeline I can introduce a logic where we check if the file exists and if not is downloaded before loading it.

File: src/oge/helpers.py
Comment (rouille) [reply src/oge/helpers.py:309]: Most values have only digit after the decimal point. The ones that don't have one digit after the decimal point followed by 9s, e.g., 3.5999999999999999

File: src/oge/helpers.py
Comment (rouille) [reply src/oge/helpers.py:184]: I did compare for 2022 the plant static attributes files before implementing the code in this PR and after:
```
>>> old[["plant_id_eia", "ba_code"]].equals(new[["plant_id_eia", "ba_code"]])
True
>>> old[["plant_id_eia", "ba_code", "state"]].compare(new[["plant_id_eia", "ba_code", "state"]])
      state      
       self other
797     NaN    WI
1924    NaN    VT
4319    NaN    MN
11630   NaN    MD
```
We can see that `ba_code` did not change between the two versions and that the 4 rows were `state` was missing are now filled out.

Comment (grgmiller) [reply src/oge/helpers.py:184]: sounds good

File: src/oge/helpers.py
Comment (grgmiller) [reply src/oge/helpers.py:415]: Thanks, this is fine as is then!

File: src/oge/helpers.py
Comment (grgmiller) [reply src/oge/helpers.py:309]: ok thanks, this makes sense then

File: src/oge/helpers.py
Comment (grgmiller) [reply src/oge/helpers.py:305]: Maybe it doesn't matter but it's just not a pattern that we use in the rest of the repo. Either way is fine. It looks like there shouldn't be any major performance difference between the two based on some quick googling

File: src/oge/helpers.py
Comment (grgmiller) [reply src/oge/helpers.py:208]: No, this is fine. I generally like using our built in function since we have it, but I don't think that there's really a benefit to using it here.

File: src/oge/helpers.py
Comment (grgmiller) [reply src/oge/helpers.py:220]: Probably no need here unless you really want to

**Review by grgmiller — APPROVED** [review]
I'm going to go ahead and approve assuming that you will be making the changes discussed in the previous review just so I don't hold up your progress.

**Comment by rouille** [comment]
> I'm going to go ahead and approve assuming that you will be making the changes discussed in the previous review just so I don't hold up your progress. 

Thanks. I launched the pipeline overnight to make sure the changes did not break it. Il will check the result when I am at my desk and commit my changes if it looks good.

File: src/oge/helpers.py
Comment (rouille) [reply src/oge/helpers.py:409]: Done

Screenshot 2024-05-22 at 9 46 54 AM

File: src/oge/column_checks.py
Comment (rouille) [reply src/oge/column_checks.py]: Done

File: src/oge/column_checks.py
Comment (rouille) [reply src/oge/column_checks.py]: Done

File: src/oge/data_pipeline.py
Comment (rouille) [reply src/oge/data_pipeline.py]: Done

File: src/oge/helpers.py
Comment (rouille) [reply src/oge/helpers.py:220]: It can be done in a quick PR later. We have bigger fishes to fry right now.

File: src/oge/helpers.py
Comment (rouille) [reply src/oge/helpers.py]: Done

File: src/oge/helpers.py
Comment (rouille) [reply src/oge/helpers.py]: Changed.

**Pull request merged**
